### DISCUSSION TITLE
Submit button not rendering and HTML null value

### DISCUSSION THREAD

**Discussion by rickeydute** [solspace/craft-freeform#1201] [Questions & Answers] [answered]
We recently updated from Freeform V4 to V5. Before the Submit button was a field but now it is added automatically, but none of our forms shows the submit button at all. 
Also, the HTML field shows a null value even though it should be pulling the proper data through the GraphQL query. 

CraftCMS: V4.8.7
Freeform: V5.1.9 Pro

**Comment by rickeydute** [comment]
Is there something I need to do for my question to be answered? Our company has an update pending waiting on this and the update to V5 has broken all of our forms and this issue is holding the update to be applied to our CMS. 
I understand that my question is vague, but I do not know what is it that you guys need to understand our issue. 
Thanks in advanced.

**Comment by kjmartens** [comment]
Hi @rickeydute,

I'm very sorry for the delay...

Have you seen the new `FreeformButtonsInterface` and `FreeformButtonsAttributesInterface` in the documentation? 🙂 

- https://docs.solspace.com/craft/freeform/v5/headless/graphql/#buttons-interface
- https://docs.solspace.com/craft/freeform/v5/headless/graphql/#buttons-attributes-interface

I will check with a developer about the HTML field issue you mention...

**Reply by rickeydute** [reply]
Hi @kjmartens 
I did read all of the docs. From my understanding, they should just work. I did look into GraphiQL to see if I needed to bring them through the GraphQL, but none of the props in the explorer show any data within the button fields. Everything comes up as null. 

As for the HTML, it is very odd. The React Dev Tools show the following: 
`{
  "id": "935267",
  "typeHandle": "form",
  "formControl": {
    "id": 5,
    "name": "Newsletter",
    "uid": "7f126de1-4717-4ee8-9ed7-0ebd81f13280",
    "hash": "6Eagq0Pxb-jMEdPld6P-OAK7W6Pwx1uiba4Luh3rACUx0SRX40UIhP5BNUHv",
    "handle": "newsletterPage",
    "returnUrl": "",
    "description": "",
    "csrfToken": "{name: \"CRAFT_CSRF_TOKEN\", value: \"sXEebM5nlqbxdShY…}",
    "honeypot": null,
    "fields": [
      {
        "handle": "html",
        "label": "HTML",
        "instructions": "",
        "type": "html",
        "required": false,
        "value": null
      },
      {
        "handle": "email",
        "label": "Email",
        "instructions": "",
        "type": "email",
        "required": false
      }
    ]
  },
  "hideLegend": true,
  "fullWidthForm": true,
  "formClass": "utils_p_0__j2yGK utils_mt_4__ps_Oe"
}`

It knows there is an HTML field, but the value comes up as null. This started happening after the update. Before it was working as intended.

**Reply by kjmartens** [reply]
Thanks for the additional information. 🙂 

Can I ask which version of Freeform 4 you updated from?

I have a developer investigating this and you should likely receive a response/solution on Monday. 🙂

**Reply by rickeydute** [reply]
@kjmartens
We upgraded from Freeform Pro 4.1.13.

**Comment by seandelaney** [comment, marked as answer]
I have fixed HTML and Rich text fields show null for value. I have also added a content property too.

No changes to buttons as I can see properties as expected.

Once my fix is merged, you will be able to see all properties for buttons, HTML and Rich Text field types using this query example.

```
query MyQuery {
  freeform {
    form(handle: "graphql") {
      pages {
        buttons {
          back
          backLabel
          layout
          save
          saveLabel
          submitLabel
        }
        rows {
          fields {
            ... on FreeformField_Html {
              id
              value
              type
              twig
              required
              pageIndex
              label
              instructions
              handle
              attributes {
                label {
                  attribute
                  value
                }
                instructions {
                  value
                  attribute
                }
                input {
                  attribute
                  value
                }
                error {
                  attribute
                  value
                }
                container {
                  value
                  attribute
                }
              }
              content
            }
            ... on FreeformField_RichText {
              id
              content
              handle
              instructions
              label
              pageIndex
              required
              type
              value
            }
          }
        }
      }
    }
  }
}
```

Output:

```
{
  "data": {
    "freeform": {
      "form": {
        "pages": [
          {
            "buttons": {
              "back": true,
              "backLabel": "Back",
              "layout": "save back|submit",
              "save": false,
              "saveLabel": "Save",
              "submitLabel": "Submit"
            },
            "rows": [
              {
                "fields": [
                  {
                    "id": 1,
                    "value": "<div class=\"flex flex-col items-center justify-center w-full\">\n    <h2>Heading</h2>\n    <p>Lorem Ipsum is simply dummy text of the printing and typesetting industry.</p>\n</div>",
                    "type": "html",
                    "twig": false,
                    "required": false,
                    "pageIndex": 0,
                    "label": "HTML",
                    "instructions": "",
                    "handle": "html",
                    "attributes": {
                      "label": [],
                      "instructions": [],
                      "input": [],
                      "error": [],
                      "container": [
                        {
                          "value": "html",
                          "attribute": "data-field-container"
                        },
                        {
                          "value": "html",
                          "attribute": "data-field-type"
                        }
                      ]
                    },
                    "content": "<div class=\"flex flex-col items-center justify-center w-full\">\n    <h2>Heading</h2>\n    <p>Lorem Ipsum is simply dummy text of the printing and typesetting industry.</p>\n</div>"
                  }
                ]
              },
              {
                "fields": [
                  {
                    "id": 2,
                    "content": "<p>My name is <strong>Sean</strong> Delaney</p>",
                    "handle": "richText",
                    "instructions": "",
                    "label": "Rich Text",
                    "pageIndex": 0,
                    "required": false,
                    "type": "rich-text",
                    "value": "<p>My name is <strong>Sean</strong> Delaney</p>"
                  }
                ]
              }
            ]
          }
        ]
      }
    }
  }
}
```

**Reply by rickeydute** [reply]
Awesome, @seandelaney! Thanks for this! I let our team know about this and will be looking forward to your merge to test on our side. 

Is there anything I can provide you guys from our code that could help troubleshoot the issue? Like our RenderForm component code?

**Reply by kjmartens** [reply]
I think I might see the confusion now. If you're referring to getting the Submit buttons to work, make sure you're accessing them this way (under `forms` -> `pages` -> `buttons`):

``` gql
query MyQuery {
  freeform {
    form(handle: "graphql") {
      pages {
        buttons {
          back
          backLabel
          layout
          save
          saveLabel
          submitLabel
        }
...
```

<img width="441" alt="Screenshot 2024-04-23 at 1 42 30 PM" src="https://github.com/solspace/craft-freeform/assets/1821674/6c874a41-02cc-4249-8930-c2486b5c1290">

**NOT** the old Freeform 4 way, under `fields` (it displaying here is an issue and we'll remove it):

``` gql
query MyQuery {
  freeform {
    form(handle: "myTestForm") {
      pages {
        rows {
          fields {
            ... on FreeformField_Submit {
              id
            }
            ... on FreeformField_Save {
              id
            }
          }
        }
      }
...
```

<img width="547" alt="Screenshot 2024-04-23 at 2 02 40 PM" src="https://github.com/solspace/craft-freeform/assets/1821674/f34b693d-d4a3-482d-9d5e-315a53abd013">

**Reply by rickeydute** [reply]
@kjmartens 
![image](https://github.com/solspace/craft-freeform/assets/11202989/ea6c91e6-cbeb-47d2-bfb0-58f04415d4fe)
`query MyQuery {
  entry(section: "pages", slug: "newsletter") {
    id
  }
  freeform {
    form(id: 9) {
      id
      handle
    }
    forms {
      id
      fields {
        ... on FreeformField_Submit {
          id
          label
        }
      }
    }
  }
}`

I have those fields in the Query, but the response does not show them. 
`{
  "data": {
    "entry": null,
    "freeform": {
      "form": {
        "id": 9,
        "handle": "umbrellasPartsRequestForm"
      },
      "forms": [
        {
          "id": 1,
          "fields": [
            {},
            {},
            {},
            {},
            {},
            {}
          ]
        },
        {
          "id": 2,
          "fields": [
            {},
            {},
            {},
            {},
            {},
            {},
            {},
            {},
            {}
          ]
        },
        {
          "id": 3,
          "fields": [
            {},
            {}
          ]
        },
        {
          "id": 4,
          "fields": [
            {},
            {},
            {},
            {},
            {},
            {},
            {},
            {},
            {}
          ]
        },
        {
          "id": 5,
          "fields": [
            {},
            {}
          ]
        },
        {
          "id": 6,
          "fields": [
            {},
            {},
            {}
          ]
        },
        {
          "id": 7,
          "fields": [
            {},
            {},
            {},
            {},
            {},
            {},
            {},
            {},
            {},
            {},
            {},
            {},
            {},
            {}
          ]
        },
        {
          "id": 9,
          "fields": [
            {},
            {},
            {},
            {},
            {},
            {},
            {},
            {},
            {},
            {},
            {},
            {},
            {},
            {},
            {},
            {},
            {}
          ]
        },
        {
          "id": 12,
          "fields": [
            {},
            {},
            {},
            {},
            {},
            {},
            {},
            {},
            {},
            {}
          ]
        },
        {
          "id": 13,
          "fields": []
        },
        {
          "id": 14,
          "fields": []
        }
      ]
    }
  }
}`

You can add any other field and the handle/label appear in the response. Buttons do not.

**Reply by seandelaney** [reply]
If you use the correct query as per [this comment](https://github.com/solspace/craft-freeform/discussions/1201#discussioncomment-9204585), it will work. I see you are still using the wrong query.

**Reply by rickeydute** [reply]
@seandelaney @kjmartens 

Sorry, I did use the wrong query. 
After spending yesterday the entire day  looking into this and speaking with a colleague, we found this: 
`export default function GetForm(handle){
    return query GetForm {
                freeform {
                    form(handle: "${handle}") {
                        name
                        uid
                        handle
                        id
                        returnUrl
                        description
                        honeypot {
                            name
                            value
                        }
                        pages {
                            buttons {
                                submitLabel
                                saveLabel
                                save
                                layout
                                backLabel
                                back
                            }
                        }
                    }
                }
            }
}`
(removed the ticks for formatting purposes)
This is the proper query that you pointed out, correct? We did update it with your suggestions, so thank you for that. 
Unfortunately, it seems like this query is incompatible with our setup. The query above is not used anymore in our system instead we query the fields of the form directly to render them within our Form block component. Since the buttons are not fields anymore, then they are missing. 
To try and fix this, we hardcoded the submit button, but currently, it does not work. Also, hardcoding the button pretty much prevents us from using custom text on the button. My colleague and I are looking into updating some other code to make this button work, but I ask you, is this the only way we have to make these buttons work? Do we have no other alternative to make these buttons render using Freeform?
